### ISSUE TITLE
Junit formatter is not outputting according to XSD Schema

### ISSUE THREAD

**Issue by CasperJ** [stoplightio/spectral#607]
**Describe the bug**
When using the JUnit formatter (`--format=junit --output=test.xml`) then the output XML is not aligned with the XSD Schema (https://github.com/windyroad/JUnit-Schema/blob/master/JUnit.xsd).

JUnit only supports "error" or "failure", not "warning", "hint" etc.

**To Reproduce**

1. Given an OpenAPI document that doesn't have an `operationId` should create a `warning`
2. Run this CLI command `lint myopapi.yaml --format=junit --output=test.xml`
3. See test.xml
```xml
<?xml version="1.0" encoding="utf-8"?>
<testsuites>
    <testsuite package="org.spectral" time="0" tests="2" errors="2" name="openapi.yaml">
        <testcase time="0" name="org.spectral.operation-description" classname="openapi">
            <warning message="Operation `description` must be present and non-empty string."><![CDATA[line 13, col 9, Warning - Operation `description` must be present and non-empty string. (operation-description)]]></warning>
        </testcase>
        <testcase time="0" name="org.spectral.operation-operationId" classname="openapi">
            <warning message="Operation should have an `operationId`."><![CDATA[line 13, col 9, Warning - Operation should have an `operationId`. (operation-operationId)]]></warning>
        </testcase>
    </testsuite>
</testsuites>
``` 

**Expected behavior**
 - testsuites/testsuite/testcase/warning is not a valid element. must be either "error" or "failure"
 - testsuites/testsuite@tests should not be equal "2" - ti should be the total number of tests run
 - if all tests succeeds then a file should still be outputted (which should include the number of tests passed)

**Additional context**
I would love to work on fixing this bug, but for now its here for reference ;)

**Comment by P0lip** [comment]
Hey @CasperJ.
Thanks for raising the issue!
As you can see in the [PR](https://github.com/stoplightio/spectral/pull/519?cf_lbyyhhwhyjj5l3rs65cb3w=t77k9m2uwbiovv34a44xh#discussion_r322863637) implementing JUnit formatter, we initially generated `error` tag for each messages regardless of their severity.
Unfortunately, doing so didn't allow us to indicate the actual severity, so decided to make that change that lead to error/warning/information/hint tags being output.

I'm more than happy to adjust it to be more spec-compliant, but would still like to preserve the severity - which is pretty important in case of Spectral.
Is there any attribute we could make use of to retain the severity in the final output?

**Comment by P0lip** [comment]
Would also love to include `path` somehow, as pointed out by @kylesykes in the following comment https://github.com/stoplightio/spectral/issues/572#issuecomment-535962788.

If you'd like to work on any of these JUnit improvements, you are very much welcome to do so :)
Let me know if I can help you in any way if you decide to pick that ticket up.

**Comment by CasperJ** [comment]
Hi @P0lip 

There is a "@type" attribute that you can use for this:
https://github.com/windyroad/JUnit-Schema/blob/49e95a79cc0bfba7961aaf779710a43a4d3f96bd/JUnit.xsd#L86

It makes sense for error/warning, but hint/info would be strange to output as errors.
As I think more about it, maybe only errors should be outputted in the file?

I mean, if you want your build system (or whatever is parsing the xml file) to fail on a warning then you should just adjust the ruleset, right?

**Comment by philsturgeon** [comment]
@P0lip although it would be nice to support all types of output, if JUnit does not have any way for us to send info and hint then we cannot do it. We need to output valid content regardless. 

@CasperJ I don’t know much about JUnit, maybe we can use error/warning and either squeeze everything into failures?

We have a new feature which allows people to control what constitutes a failure and what results should be returned: https://github.com/stoplightio/spectral/pull/540 

If we can let this setting control what shows up in JUnit like any other format then that would be less confusing. That said JUnit can just throw out hint and info if it has nowhere to put them, it just might lead to confusion as a user could have the CLI set to fail on info but then not see anything. Maybe that’s on them.

**Comment by CasperJ** [comment]
I think we need to understand that there is a difference between formatting for console and JUnit/xUnit/mstest etc.

Humans look at console and that can show hints and other good recommendations.
JUnit is usually read by machines aka. build systems which uses it to decide if it should fail or pass a build. So in this context hints doesn't make sense as it would know what to do with it :)

In JUnit, if you pass the testsuites then it would "just" show "green" 123/123 tests passed. (That's pretty much whats in the XML)

In case of a failure it would report "red" 122/123 tests passed and then the details of the error (incl. stacktrace, tracing etc. if present)

So what I'm trying to say is that what goes to the console vs. to a file doesn't need to have the same information in it.

xUnit has the same construct in their schema: https://xunit.net/docs/format-xml-v2 => errors are documented - successes are only counted

**Comment by philsturgeon** [comment]
That makes sense. That’s why I’m fine with going either way, I suppose testing systems either work and it’s a +1 or they fail and that’s a problem which has more details.

I think we should respect the fail severity and let folks decide if error or warning is what should be shown or trigger a failure, and the format itself can just trim out anything lower as they’re not relevant to the format. 

--
Phil Sturgeon
@philsturgeon
